### PR TITLE
Download enumerator refactor

### PR DESCRIPTION
**Pull request by adreed-msft**
Pretty roughly based off the upload enumerator refactor. There's only a couple of places that really needed altering.

### CODE REVIEW

File: cmd/copy.go [hunk 1]
```diff
@@ -1246,6 +1250,7 @@ func init() {
 	rootCmd.AddCommand(cpCmd)
 
 	// filters change which files get transferred
+	cpCmd.PersistentFlags().BoolVar(&raw.copyContents, "strip-top-dir", false, "copy everything under the source, not the source itself")
```
Comment (zezha-msft) [inline cmd/copy.go]: >copy everything under the source, not the source itself [](start = 77, length = 55)

consider rephrase it be more user-friendly. "not the source itself" => something like "the top directory itself will not be replicated on the destination side"

File: cmd/copy.go [hunk 2]
```diff
@@ -83,6 +83,7 @@ type rawCopyCmdArgs struct {
 	// filters from flags
 	listOfFilesToCopy string
 	recursive         bool
+	copyContents      bool
```
Comment (zezha-msft) [inline cmd/copy.go]: >copyContents [](start = 1, length = 12)

rename to correspond to the flag name.

This whole command "copies contents" so the field name can be quite confusing.

File: cmd/copyDownloadBlobEnumerator.go [hunk 3]
```diff
@@ -1,341 +0,0 @@
...
-
-	"strings"
-
-	"github.com/Azure/azure-storage-azcopy/common"
-	"github.com/Azure/azure-storage-azcopy/ste"
-	"github.com/Azure/azure-storage-blob-go/azblob"
-)
-
-type copyDownloadBlobEnumerator common.CopyJobPartOrderRequest
-
```
Comment (zezha-msft) [inline cmd/copyDownloadBlobEnumerator.go]: aw yiiiis

**Comment by zezha-msft** [comment]
Looking good! Now we are really seeing the advantage of the modular design:
image

**Comment by zezha-msft** [comment]
Please consider adding the changelog entry now, to explain the change in behavior for download, to avoid forgetting the details at release time

File: cmd/copyDownloadBlobEnumerator.go (hunk 3 above)
Comment (zezha-msft) [reply cmd/copyDownloadBlobEnumerator.go]: image

File: testSuite/scripts/test_upload_block_blob.py [hunk 4]
```diff
@@ -504,15 +504,15 @@ def test_download_blob_include_flag(self):
             x = json.loads(result, object_hook=lambda d: namedtuple('X', d.keys())(*d.values()))
         except:
             self.fail('error parsing the output in Json Format')
-        self.assertEquals(x.TransfersCompleted, 3)
+        self.assertEquals(x.TransfersCompleted, 6)
```
Comment (JohnRusk) [inline testSuite/scripts/test_upload_block_blob.py]: It seems weird that we didn't change the test at all, and the test only includes 3 files with the include pattern, but now we are downloading 6 files instead of 3. What's up there?

File: testSuite/scripts/test_upload_block_blob.py [hunk 5]
```diff
@@ -559,14 +559,14 @@ def test_download_blob_exclude_flag(self):
         except:
             self.fail('error parsing the output in JSON Format')
         # Number of expected successful transfer should be 18 since two files in directory are set to exclude
-        self.assertEquals(x.TransfersCompleted, 17)
+        self.assertEquals(x.TransfersCompleted, 14)
```
Comment (JohnRusk) [inline testSuite/scripts/test_upload_block_blob.py]: Ditto, why the change in behaviour with so little change to the test?

File: cmd/copy.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (adreed-msft) [reply cmd/copy.go]: stripTopDir is inferred when wildcarding a local path. Otherwise, yes it is supported with a local source. Because stripTopDir defaults to false, error checking an inconsistency isn't possible because we don't know whether it was _explicitly set_ to false or _defaulted_ to false.

If we wanted to (pointlessly in my opinion) turn it into an enum of (default, false, true), we could manipulate that functionality. However, if a user really wanted the result of --strip-top-dir=false with a wildcard, couldn't they just manually append the top directory to the destination themself?

File: cmd/copyEnumeratorInit.go [hunk 6]
```diff
@@ -40,7 +52,7 @@ func (cca *cookedCopyCmdArgs) initEnumerator(jobPartOrder common.CopyJobPartOrde
 
 	// Ensure we're only copying from a directory with a trailing wildcard or recursive.
 	isSourceDir := traverser.isDirectory(true)
-	if isSourceDir && !cca.recursive {
+	if isSourceDir && !cca.recursive && !cca.copyContents {
```
[the remark this answers is not in the document]
Comment (adreed-msft) [reply cmd/copyEnumeratorInit.go]: So, back when this was `copy-contents` I'd imagined trying to use it _without_ recursive. In this case, it'd have non-recursive sync-like functionality (sans the LMT checking etc.) where it just copies the contents of the directory but not sub-directories.

File: cmd/zc_traverser_file.go [hunk 7]
```diff
@@ -101,24 +101,22 @@ func (t *fileTraverser) traverse(processor objectProcessor, filters []objectFilt
 			for _, fileInfo := range lResp.FileItems {
 				f := currentDirURL.NewFileURL(fileInfo.Name)
 
-				//// TODO: the cost is high while otherwise we cannot get the last modified time. As Azure file's PM description, list might get more valuable file properties later, optimize the logic after the change...
-				//// TODO this traverser is only being used by rm at the moment, so we don't need the properties, uncomment in the future when this is no longer true
-				//fileProperties, err := f.GetProperties(t.ctx)
-				//if err != nil {
-				//	return err
-				//}
+				fileProperties, err := f.GetProperties(t.ctx)
```
[the remark this answers is not in the document]
Comment (adreed-msft) [reply cmd/zc_traverser_file.go]: This is false. The file enumerator was only used for remove.

The problem I encountered is that with this disabled, I'd duplicate the source to the destination as empty files.

[the remark this answers is not in the document]
Comment (adreed-msft) [reply cmd/zc_traverser_file.go]: Remove didn't actually need much info to remove the file, hence why we never supplied that information.

We _could_ get the info in STE, yes. Though, files sync is something we intend to do and this info is needed for sync regardless.

File: cmd/zc_traverser_file.go [hunk 8]
```diff
@@ -101,24 +101,22 @@ func (t *fileTraverser) traverse(processor objectProcessor, filters []objectFilt
...
 				relativePath := strings.TrimPrefix(fileURLParts.DirectoryOrFilePath, targetURLParts.DirectoryOrFilePath)
 				relativePath = strings.TrimPrefix(relativePath, common.AZCOPY_PATH_SEPARATOR_STRING)
 
 				storedObject := storedObject{
-					name:         getObjectNameOnly(fileInfo.Name),
-					relativePath: relativePath,
-					//lastModifiedTime: fileProperties.LastModified(),
-					//md5:              fileProperties.ContentMD5(),
-					//size:             fileProperties.ContentLength(),
+					name:             getObjectNameOnly(fileInfo.Name),
```
[the remark this answers is not in the document]
Comment (adreed-msft) [reply cmd/zc_traverser_file.go]: Changing to the newStoredObject func, I will not revert the getProperties call due to
- Potential future usage in sync
- Getting these properties in STE feels somewhat janky when all other cases get these properties in the enumerator
- The only reason we didn't get properties here is because it was only `rm` that was using it.

File: testSuite/scripts/test_upload_block_blob.py (hunk 4 above)
Comment (adreed-msft) [reply testSuite/scripts/test_upload_block_blob.py]: So, include and exclude previously only did things on the root directory (because they were checking the relative path, not the file name).

We ended up declaring this a bug and that the include and exclude filters should affect file names under all sub directories.

Comment (adreed-msft) [reply testSuite/scripts/test_upload_block_blob.py]: As a result, there are 3 matching files under each directory. Therefore, 6 files get included.

File: testSuite/scripts/test_upload_block_blob.py (hunk 5 above)
Comment (adreed-msft) [reply testSuite/scripts/test_upload_block_blob.py]: Explained in above comment.

File: cmd/copyDownloadBlobEnumerator.go (hunk 3 above)
Comment (adreed-msft) [reply cmd/copyDownloadBlobEnumerator.go]: image

File: testSuite/scripts/test_blob_download.py [hunk 9]
```diff
@@ -350,19 +357,21 @@ def test_blob_download_wildcard_recursive_true_1(self):
...
+            execute_azcopy_copy_command_get_output()
         # parse the result to get the last job progress summary
         result = util.parseAzcopyOutput(result)
         try:
             # parse the Json Output
             x = json.loads(result, object_hook=lambda d: namedtuple('X', d.keys())(*d.values()))
         except:
             self.fail('error parsing the output in Json Format')
-        # since the wildcards '*/*' exists at the end of dir_name in the sas
+        # since the wildcards '*' exists at the end of dir_name in the sas
```
[the remark this answers is not in the document]
Comment (adreed-msft) [reply testSuite/scripts/test_blob_download.py]: Done

File: cmd/copyEnumeratorInit.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (JohnRusk) [reply cmd/copyEnumeratorInit.go]: So now that its strip-to-dir, what's the relationship with recursive?  E.g. only usable with recursive?

File: cmd/zc_traverser_file.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (JohnRusk) [reply cmd/zc_traverser_file.go]: Ah, I get it now.  Yes, let's say with the properties call that you are making. I had forgotten it wasn't already used for everything.

File: cmd/copy.go [hunk 10]
```diff
@@ -183,15 +183,17 @@ func (raw rawCopyCmdArgs) cook() (cookedCopyCmdArgs, error) {
...
 	// copy&transform flags to type-safety
-	cooked.copyContents = raw.copyContents
+	cooked.stripTopDir = raw.stripTopDir
 	cooked.recursive = raw.recursive
 	cooked.followSymlinks = raw.followSymlinks
 	cooked.withSnapshots = raw.withSnapshots
 	cooked.forceWrite = raw.forceWrite
 
-	// Local supports wildcards just fine.
+	// cooked.stripTopDir is effectively a workaround for the lack of wildcards in remote sources.
```
Comment (JohnRusk) [inline cmd/copy.go]: Thanks for the comment. I get it now.

File: cmd/copy.go [hunk 11]
```diff
@@ -957,7 +958,15 @@ func (cca *cookedCopyCmdArgs) processCopyJobPartOrders() (err error) {
...
-		cca.destination = cleanLocalPath(cca.destination)
+		var result string
+		result, err = filepath.Abs(cca.destination)
+
+		if err != nil {
+			return err
+		}
+
+		cca.destination = cleanLocalPath(result)
+		jobPartOrder.DestinationRoot = cca.destination
```
Comment (JohnRusk) [inline cmd/copy.go]: Minor: not needed because the same thing is done on the very next line (immediately after the switch ends)

File: cmd/copy.go [hunk 12]
```diff
@@ -1250,7 +1259,7 @@ func init() {
 	rootCmd.AddCommand(cpCmd)
 
 	// filters change which files get transferred
-	cpCmd.PersistentFlags().BoolVar(&raw.copyContents, "strip-top-dir", false, "copy everything under the source, not the source itself")
+	cpCmd.PersistentFlags().BoolVar(&raw.stripTopDir, "strip-top-dir", false, "strip the source's root folder from the destination path, akin to \"cp dir/*\" (ex. sourcedir/subdir1/file1 copies to subdir1/file1 on destination)")
```
Comment (JohnRusk) [inline cmd/copy.go]: Maybe add "whereas without strip-to-dir, it would copy to sourcedir/subdir1/file1 on the destination"

File: cmd/copyEnumeratorInit.go [hunk 13]
```diff
@@ -52,7 +52,7 @@ func (cca *cookedCopyCmdArgs) initEnumerator(jobPartOrder common.CopyJobPartOrde
 
 	// Ensure we're only copying from a directory with a trailing wildcard or recursive.
 	isSourceDir := traverser.isDirectory(true)
-	if isSourceDir && !cca.recursive && !cca.copyContents {
+	if isSourceDir && !cca.recursive && !cca.stripTopDir {
```
Comment (JohnRusk) [inline cmd/copyEnumeratorInit.go]: Please see new question in my earlier thread about this (from first-pass review)

File: cmd/copy.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (JohnRusk) [reply cmd/copy.go]: Good point re can't tell why its false.  I'm closing this comment thread.

File: cmd/copyEnumeratorInit.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (adreed-msft) [reply cmd/copyEnumeratorInit.go]: Same relationship as before, I think. @zezha-msft thoughts on retaining this functionality or making it so that it must be used alongside recursive?

[the remark this answers is not in the document]
Comment (zezha-msft) [reply cmd/copyEnumeratorInit.go]: `--strip-top-dir` has the meaning of targeting the given path's children instead of itself, in that sense I think the current behavior (target only the direct children but not sub-dirs).

However, the name of the flag may create confusion, so the behavior should be clearly documented and shown in examples.

[the remark this answers is not in the document]
Comment (adreed-msft) [reply cmd/copyEnumeratorInit.go]: Alright, document it I will.

File: cmd/copy.go [hunk 14]
```diff
@@ -1258,7 +1258,7 @@ func init() {
 	rootCmd.AddCommand(cpCmd)
 
 	// filters change which files get transferred
-	cpCmd.PersistentFlags().BoolVar(&raw.stripTopDir, "strip-top-dir", false, "strip the source's root folder from the destination path, akin to \"cp dir/*\" (ex. sourcedir/subdir1/file1 copies to subdir1/file1 on destination; whereas without --strip-top-dir, it copies to sourcedir/subdir1/file1)")
+	cpCmd.PersistentFlags().BoolVar(&raw.stripTopDir, "strip-top-dir", false, "strip the source's root folder from the destination path, akin to \"cp dir/*\". Without --recursive, just copies files under the folder but does not recurse into sub-directories akin to a non-recursive sync (ex. sourcedir/subdir1/file1 copies to subdir1/file1 on destination; whereas without --strip-top-dir, it copies to sourcedir/subdir1/file1)")
```
Comment (JohnRusk) [inline cmd/copy.go]: Can you re-order and tweak wording slightly please, to:

> strip the source's root folder from the destination path, akin to \"cp dir/*\".  E.g. sourcedir/subdir1/file1 copies to subdir1/file1 on destination; whereas without --strip-top-dir, it copies to sourcedir/subdir1/file1.  May be used with and without --recursive. Without --recursive, just copies files under the folder but does not recurse into sub-directories

Reason for suggested change is that the recursive stuff is detail, and having it before the main example would confuse users.  And the allusion to "non-recursive sync" potentially added questions ("Oh, I need to learn what sync does before I can understand this flag" etc)

**Comment by adreed-msft** [comment]
Squash and merge, squash and merge!
